src/libtls/tls_hkdf.c

```diff
@@ -0,0 +1,602 @@
+	 * Length of this hash output in bytes.
+	 */
+	size_t hash_length;
+
+	/**
+	 * Preshared key IKM
+	 */
+	chunk_t *psk;
+
+	/**
+	 * (EC)DHE IKM
+	 */
+	chunk_t *shared_secret;
+
+	/**
+	 * Pseudorandom function used.
+	 */
+	prf_t *prf;
+
+	/**
+	 * Pseudorandom plus function used.
+	 */
+	prf_plus_t *prf_plus;
+
+	/**
+	 * Hasher used.
+	 */
+	hasher_t *hasher;
+
+	/**
+	 * Salt used.
+	 */
+	chunk_t salt;
+
+	/**
+	 * IKM used.
+	 */
+	chunk_t ikm;
+
+	/**
+	 * PRK used.
+	 */
+	chunk_t prk;
+
+	/**
+	 * Info used.
+	 */
+	chunk_t info;
+
+	/**
+	 * OKM used.
+	 */
+	chunk_t okm;
+
+	/**
+	 * Length of the desired key output.
+	 */
+	size_t L;
+};
+
+/* HKDF-Extract(this->salt, this->ikm) -> this->prk */
+static bool extract(private_tls_hkdf_t *this)
+{
+	if (!this->prf->set_key(this->prf, this->salt))
+	{
+		DBG2(DBG_TLS, "unable to set PRF salt");
+		return FALSE;
+	}
+	if(!this->prf->allocate_bytes(this->prf, this->ikm, &this->prk))
+	{
+		DBG2(DBG_TLS, "unable to allocate PRF space");
+		return FALSE;
+	}
+
+	DBG1(DBG_TLS, "PRK: %B", &this->prk);
+
+	return TRUE;
+}
+
+/* HKDF-Expand(this->prk, this->info, this->L) -> this->okm */
+static bool expand(private_tls_hkdf_t *this)
+{
+	if (!this->prf->set_key(this->prf, this->prk))
+	{
+		DBG2(DBG_TLS, "unable to set PRF PRK");
+		return FALSE;
+	}
+	this->prf_plus = prf_plus_create(this->prf, TRUE, this->info);
+	if (!this->prf_plus->allocate_bytes(this->prf_plus, this->L, &this->okm))
+	{
+		DBG2(DBG_TLS, "unable to allocate PRF plus space");
+		return FALSE;
+	}
+
+	DBG1(DBG_TLS, "OKM: %B", &this->okm);
+
+	return TRUE;
+}
+
+static bool expand_label(private_tls_hkdf_t *this, chunk_t *label, chunk_t *context, uint16_t length)
+{
+	/*
+	 * pack label and context into structure hkdflabel
+	 */
+	if (label->len < 7 || label->len > 255 ||
+		context->len < 0 || context->len > 255)
+	{
+		return FALSE;
+	}
+	/* HKDFLabel */
+	size_t offset = 0;
+	size_t chunk_len = 2 + 1 + label->len + 1 + context->len;
+	this->info = chunk_create(malloc(chunk_len), chunk_len);
+	this->info.ptr[1] = length & 0xFF;
+	this->info.ptr[0] = length >> 8;
+	this->info.ptr[2] = label->len;
+	offset += 3;
+	memcpy(&(this->info).ptr[offset], label->ptr, label->len);
+	offset += label->len;
+	this->info.ptr[offset] = context->len;
+	offset += 1;
+	memcpy(&(this->info).ptr[offset], context->ptr, context->len);
+	offset += context->len;
+
+	expand(this);
+
+	return TRUE;
+}
+
+static bool derive_secret(private_tls_hkdf_t *this, chunk_t *label, chunk_t *messages)
+{
+	/*
+	 * calculate hash of message
+	 */
+	/*
+	 * Hash messages and call expand
+	 */
+	size_t chunk_len = 32;
+	chunk_t context = chunk_create(malloc(chunk_len), chunk_len);
+
+	if (!this->hasher || !this->hasher->allocate_hash(this->hasher, *messages, &context))
+	{
+		DBG2(DBG_TLS, "%N not supported", hash_algorithm_names, this->hash_algorithm);
+		return FALSE;
+	}
+
+	expand_label(this, label, &context, 32);
+	chunk_free(&context);
+
+	return TRUE;
+}
+
```

Comment:
I'm having doubts about the use of specific functions that announce a move. The phases in hkdf are similar to a state machine (phase = state), thus it might be a good idea to use a switch/case like in other files (cf. tls_peer.c), which consist of phase 0-4, rather than using specific methods that announce the change of a phase.
It would also reduce the move-functions to just one, something like `check_phase()` that contains the state machine's phases.